encoding: UnsafeConvertStringToBytes panic on large input

Panic importing 2.7GB file https://cockroachlabs.slack.com/archives/CV581CE78/p1696286878568879
```
panic: runtime error: slice bounds out of range [::5626539592] with length 2147483647 [recovered]
	panic: runtime error: slice bounds out of range [::5626539592] with length 2147483647

goroutine 933315 [running]:
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0x14013faa600, {0x109afcec0, 0x14005d2f980}, {0x109722b60?, 0x14011e72000?})
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1199 +0x164
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1({0x109afcec0, 0x14005d2f980}, {0x0?})
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:952 +0x4c
panic({0x109722b60, 0x14011e72000})
	GOROOT/src/runtime/panic.go:884 +0x204
github.com/cockroachdb/cockroach/pkg/util/encoding.UnsafeConvertStringToBytes(...)
	github.com/cockroachdb/cockroach/pkg/util/encoding/encoding.go:927
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*DBytes).UnsafeBytes(...)
	github.com/cockroachdb/cockroach/pkg/sql/sem/tree/datum.go:1674
github.com/cockroachdb/cockroach/pkg/sql.(*fileUploadMachine).writeFile(0x14008e3e810, {0x109afce18, 0x1400a0c8dc0}, 0x1?)
	github.com/cockroachdb/cockroach/pkg/sql/copy_file_upload.go:204 +0x210
github.com/cockroachdb/cockroach/pkg/sql.(*copyMachine).processCopyData(0x14009874e00, {0x109afce18, 0x1400a0c8dc0}, {0x0, 0x0}, 0x1)
	github.com/cockroachdb/cockroach/pkg/sql/copy_from.go:671 +0x3f4
```

Jira issue: CRDB-31994

--- a/pkg/util/encoding/BUILD.bazel
+++ b/pkg/util/encoding/BUILD.bazel
@@ -46,6 +46,7 @@ go_test(
         "//pkg/geo/geographiclib",
         "//pkg/geo/geopb",
         "//pkg/roachpb",
+        "//pkg/testutils/skip",
         "//pkg/util/bitarray",
         "//pkg/util/duration",
         "//pkg/util/ipaddr",

--- a/pkg/util/encoding/encoding.go
+++ b/pkg/util/encoding/encoding.go
@@ -17,7 +17,6 @@ import (
 	"encoding/hex"
 	"fmt"
 	"math"
-	"reflect"
 	"strconv"
 	"strings"
 	"time"
@@ -913,18 +912,12 @@ func prettyPrintInvertedIndexKey(b []byte) (string, []byte, error) {
 // modified if the input string is expected to be used again - doing so could
 // violate Go semantics.
 func UnsafeConvertStringToBytes(s string) []byte {
+	// unsafe.StringData output is unspecified for empty string input so always
+	// return nil.
 	if len(s) == 0 {
 		return nil
 	}
-	// We unsafely convert the string to a []byte to avoid the
-	// usual allocation when converting to a []byte. This is
-	// kosher because we know that EncodeBytes{,Descending} does
-	// not keep a reference to the value it encodes. The first
-	// step is getting access to the string internals.
-	hdr := (*reflect.StringHeader)(unsafe.Pointer(&s))
-	// Next we treat the string data as a maximally sized array which we
-	// slice. This usage is safe because the pointer value remains in the string.
-	return (*[0x7fffffff]byte)(unsafe.Pointer(hdr.Data))[:len(s):len(s)]
+	return unsafe.Slice(unsafe.StringData(s), len(s))
 }
 
 // EncodeStringAscending encodes the string value using an escape-based encoding. See
@@ -985,18 +978,7 @@ func EncodeEmptyArray(b []byte) []byte {
 
 // EncodeStringDescending is the descending version of EncodeStringAscending.
 func EncodeStringDescending(b []byte, s string) []byte {
-	if len(s) == 0 {
-		return EncodeBytesDescending(b, nil)
-	}
-	// We unsafely convert the string to a []byte to avoid the
-	// usual allocation when converting to a []byte. This is
-	// kosher because we know that EncodeBytes{,Descending} does
-	// not keep a reference to the value it encodes. The first
-	// step is getting access to the string internals.
-	hdr := (*reflect.StringHeader)(unsafe.Pointer(&s))
-	// Next we treat the string data as a maximally sized array which we
-	// slice. This usage is safe because the pointer value remains in the string.
-	arg := (*[0x7fffffff]byte)(unsafe.Pointer(hdr.Data))[:len(s):len(s)]
+	arg := UnsafeConvertStringToBytes(s)
 	return EncodeBytesDescending(b, arg)
 }
 

--- a/pkg/util/encoding/encoding_test.go
+++ b/pkg/util/encoding/encoding_test.go
@@ -26,6 +26,7 @@ import (
 	// Blank import so projections are initialized correctly.
 	_ "github.com/cockroachdb/cockroach/pkg/geo/geographiclib"
 	"github.com/cockroachdb/cockroach/pkg/geo/geopb"
+	"github.com/cockroachdb/cockroach/pkg/testutils/skip"
 	"github.com/cockroachdb/cockroach/pkg/util/bitarray"
 	"github.com/cockroachdb/cockroach/pkg/util/duration"
 	"github.com/cockroachdb/cockroach/pkg/util/ipaddr"
@@ -831,7 +832,7 @@ func TestEncodeBitArray(t *testing.T) {
 				0, 0xba}},
 		{bitarray.MakeZeroBitArray(62),
 			[]byte{0x3a,
-				0x88, //word 0
+				0x88, // word 0
 				0, 0xc6}},
 		{bitarray.Not(bitarray.MakeZeroBitArray(62)),
 			[]byte{0x3a,
@@ -2620,6 +2621,36 @@ func TestPrettyPrintValueEncoded(t *testing.T) {
 		if str != test.expected {
 			t.Errorf("%d: got %q expected %q", i, str, test.expected)
 		}
+	}
+}
+
+func TestUnsafeConvertStringToBytes(t *testing.T) {
+	// Large input runs slowly.
+	skip.UnderStress(t)
+
+	testCases := []struct {
+		desc     string
+		input    string
+		expected []byte
+	}{
+		{
+			desc:     "empty",
+			input:    "",
+			expected: nil,
+		},
+		{
+			// Previous impl could not handle strings longer than math.MaxInt32.
+			// See https://github.com/cockroachdb/cockroach/issues/111626
+			desc:     "large input",
+			input:    string(make([]byte, math.MaxInt32+1)),
+			expected: make([]byte, math.MaxInt32+1),
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.desc, func(t *testing.T) {
+			actual := UnsafeConvertStringToBytes(tc.input)
+			require.Equal(t, tc.expected, actual)
+		})
 	}
 }
 

